### DISCUSSION TITLE
Quadratic Throtle

### DISCUSSION THREAD

**Discussion by Paulii6661** [danielnilsson9/bbs-fw#226] [Ideas]
ive added quadratic throttle for beter control in low currents maybe you would want to implement it in your code.

edited app.c

```c++

bool apply_throttle(uint8_t* target_current, uint8_t throttle_percent)

{
	if ((assist_level_data.level.flags & ASSIST_FLAG_THROTTLE) && throttle_percent > 0 && throttle_ok())
	{
		// Apply quadratic throttle
        uint8_t normalized_throttle = (uint8_t)((throttle_percent * throttle_percent) / 100);

		uint8_t current = (uint8_t)MAP16(normalized_throttle, 0, 100, g_config.throttle_start_percent, assist_level_data.level.max_throttle_current_percent);

		// Throttle always overrides PAS if global speed limit is configured for throttle.
		bool global_throttle_limit_active =
			g_config.throttle_global_spd_lim_percent > 0 && 
			(
				g_config.throttle_global_spd_lim_opt == THROTTLE_GLOBAL_SPEED_LIMIT_ENABLED ||
				(g_config.throttle_global_spd_lim_opt == THROTTLE_GLOBAL_SPEED_LIMIT_STD_LVLS && operation_mode == OPERATION_MODE_DEFAULT)
			);

		if (current >= *target_current || global_throttle_limit_active)
		{
			*target_current = current;	
			return true;
		}
	}

	return false;
}

**Comment by Energeek136** [comment]
Hey, can you explain what exactly is changed I cant quite understand what I need to change. Thanks

**Reply by Paulii6661** [reply]
Throttle response in stock code is linear meaning if you give throttle 20% motor current response will be 20% if you give 50% current will be 50%

I this code you read throttle % multiply it by throttle percent and divide by 100 (Apply quadratic throttle
uint8_t normalized_throttle = (uint8_t)((throttle_percent * throttle_percent) / 100);) meaning you have a lot smoother throttle response because if you give 50% then 50*50/2 is 25% current if you give 75% throttle you have approx 50% current so the response isn't linear but quadratic meaning a lot more controll and a smooth throttle ride.

**Reply by dav0000000** [reply]
I don't use a throttle but doesn't the current also get smoothed out by ramp up and down?

**Reply by Paulii6661** [reply]
No that function works only when you pedal I think... I don't have pedals so I'm only guessing.

**Reply by dav0000000** [reply]
> No that function works only when you pedal

I think you are correct there. Do you think commenting out line 783 in app.c would make throttle ramp up?

**Reply by Paulii6661** [reply]
In app.c line 783 is "		return true;" so I don't think so probably you meant other line.

But you don't want current ramp up if you use current ramp up you won't be able to accelerate fast enough when you actually need it. I recommend quadratic throttle you have very good control over throttle meaning it's smooth but when you need it you get full power just adjust code it starts at line 484 delete what's there and copy paste what I added here as first post. Try it out and let me know. After I changed to quadratic throttle I would never go back to linear.

**Reply by dav0000000** [reply]
Yeah wrong line, it was line 501 which is the same statement. But I suspect this will mess up the pas/throttle override so it would need a bit more investigation. I don't use the throttle and also don't use daniel's code. I have kept my own seperate version for the last 2.5 years but it is heavily based on daniel's code. The quadratic throttle does sound good though. I'd want to try throttle with ramp but probably with a faster ramp up than normal.

**Reply by Paulii6661** [reply]
What did you add to the code? Did you touch battery.c?
I'm using Daniel's code but modified quadratic throttle, modified cruise to activate only by twisting throttle (I don't have pedals so I wasn't able to spin them to activate cruise), I'm using pretension someone shared here long time ago. I'm really curious what changes you did to the code.

**Reply by dav0000000** [reply]
I've changed a lot and didn't implement some of the changes that other people requested if I didn't need it. I have a DPC18 with 9 pas levels, a ebrake and shift sensor. I've added a pretty accurate amp based soc counter, no pedal gear shifting, speed based gear changing, speed based start current and coasting current meaning the motor continues to run even when you stop pedalling so there is no slack in the drivetrain. These can be turned on and off and won't work if an ebrake is not detected. I have 6 modes, standard, eco, tour, turbo and climb for long hills. They can be changed at any time using PUSH. I changed cruise control so it maintains speed but it also needs an ebrake. As I ghost pedal the current is proportional to how fast I pedal so when I come to a hill I just spin the pedals faster and get more power. I keep track of a bunch of stats like max watts, temperature, distance, time and save them during power off. I log values like motor request response times which can be displayed in range or calorie fields. I created a menu type setup so I can adjust all of this from the display using PUSH in combination with the pas level. PUSH can be given more power if required but only works in PAS0. There's other stuff as well like locking the bike so the motor won't work unless unlocked. I do lvc different, instead of ramping down I keep track of the percent of current that will have the battery resting on lvc so always get max possible power without tripping the bms. There is a pretty accurate range calculation. I don't use any of this data that I collect these days so it's kinda useless but I used it while I was getting it all working. I've not published my code because I don't want to support it or have law enforcement or a thief be able to work out how to use it. It's been my favourite hobby for a long time now.

**Reply by Paulii6661** [reply]
Wow that's pretty impressive I have same build like you dpc18, ebrake, shift sensor and bbshd. But after I flashed Daniel's firmware I started to enjoy much more riding around without pedaling so I adjusted my bike accordingly.

Would you mind sharing your battery.c and your cruise code?

I wanted to implement some strategies for accurate SOC but everything failed what I came up with so I'm using stock battery.c meaning soc isn't calculating unless motor is off and that's an issue for me because sometimes I go 30 - 50mins on straight road or climbing steep hill and I don't get update especially since I use chain tensioner (you call it coasting) motor is spinning even if not under load so my SOC is so bad I can't use it I always have to rely on voltage.

With cruise I tried implement something similar like escooters like xiaomi uses when you achieve some speed and stay there for couple of seconds it will keep that speed but I failed miserably so I gave this one up and just using on off with throttle twist.

**Reply by dav0000000** [reply]
I don't think sharing battery.c will help as this is true soc counting and takes place in app.c by reading the amps from motor.c so it just counts down 1% at a time. It's really cool but you have to get your battery milliamp hours set correctly. Cruise control won't work either with just a throttle but it probably won't be hard to get it working. Is it possible for me to contact you somehow on here?

**Comment by EigenMania** [comment]
Hey I just tried this and it's great! So much easier to feather the throttle and start the bike from a stand-still, and you still get full power when you mash the throttle quickly. Thanks for the tip, and I'll definitely keep this :D

**Comment by dav0000000** [comment]
uint8_t normalized_throttle = (uint8_t)((throttle_percent * throttle_percent) / 100);

Is there an overflow issue in this code above as throttle_percent is a uint8_t.

When I compile with the line below there is an extra byte.

uint8_t normalized_throttle = (uint8_t)(((uint16_t)throttle_percent * throttle_percent) / 100);

Any thoughts, anyone?

**Reply by EigenMania** [reply]
Compiler is probably implicitly casting the product of both uint8_t types to uint16_t to avoid overflow. I think a slightly "nicer" way to do the computation is with explicit exponentiation: 

![image](https://github.com/user-attachments/assets/e699add1-0104-46b3-bf02-18d5bdba0b32)

Then you can also set the exponent to whatever you want (ie 1.5 would be less aggressive, 3 more aggressive, and if you make the power high enough you just get the Dirac delta function throttle ;) )

**Reply by EigenMania** [reply]
![image](https://github.com/user-attachments/assets/d345ccaf-4063-42be-beea-832223974545)

**Reply by dav0000000** [reply]
Ok. Thanks. That makes sense. You guys make me want to put a throttle on again.

**Reply by Paulii6661** [reply]
EigenMania nice explanation I have zero clue what are you talking about but I hope it will work for you guys same as it works for me :D

Dave absolutely I suggest you try it out. Stock throttle is rubbish I had to have 5 pas levels for throttle now I use only one 100% power and can enjoy the ride very much. :)

**Reply by dav0000000** [reply]
@EigenMania pow function wasn't found on my mac or windows machines but it did find powf in math.h. Hoping this is the same thing.

**Reply by EigenMania** [reply]
pow() uses double, and powf() uses float. So pow() is only really necessary if you need the extra precision (64-bit vs 32-bit). powf() will be fine for these computations.

**Comment by silenec** [comment]
I flashed this yesterday and it is really nice, thanks for the code. Now I can pre-tension the drivetrain smoothly with just the throttle. I feel like I get lower max motor power now and that makes sense because my Throttle End Voltage (mV) is probably wrong and quadratic throttle makes this even more apparent. I always had it set to 4200mV. What value do you guys use with the stock BBS02 thumb throttle? Has anyone actually measured the throttle voltage range? Wiki says 3600mV, is that correct? What is the highest end voltage that still gives you 100% power?

**Reply by dav0000000** [reply]
The Bafang thumb throttle in my kit is 820mV resting then 898-4277mV. I set 900-4250. Ali express half twist throttle is 839mV resting then 858-4199mV. I use 900-4200.

**Reply by silenec** [reply]
Interesting, thanks. I wonder why it's slower then.
I use Max Throttle Current (%) 80
Linear throttle gave me 80% current with full throttle (I think). How does this work with quadratic throttle, do I still get ((100x100/100)x0,8) = 80% ? The power almost feels like ((100x0,8)x(100x0,8)/100) = 64% but sadly I don't understand code so can't verify.

**Reply by Paulii6661** [reply]
Hi Šílenec, code is only calculating throttle % dosent have to do anything with current meaning it calculates throttle percent * throttle percent /100 to get it back to 0 - 100% range. Meaning if you have your throttle set up correctly then you have at 100% 80% of current. Also please note that every throttle is slightly different and using it overtime also makes a little changes at least from my experience.
I suggest you set it to 100% and use stock settings for throttle which is 3600 upper limit if you are unsure.

**Reply by Blingbug** [reply]
> Hi Šílenec, code is only calculating throttle % dosent have to do anything with current meaning it calculates throttle percent * throttle percent /100 to get it back to 0 - 100% range. Meaning if you have your throttle set up correctly then you have at 100% 80% of current. Also please note that every throttle is slightly different and using it overtime also makes a little changes at least from my experience. I suggest you set it to 100% and use stock settings for throttle which is 3600 upper limit if you are unsure.

can confirm, my ali half-twist caps at ~4000, so variations can be great

**Comment by kklmm** [comment]
Hi @danielnilsson9, i see you added code for quadratic toggle to firmware 1.5.0 but forgot to add it to config tool.

**Reply by silenec** [reply]
It says "Add compile time firmware config for throttle curve map". I think that means you can't change this setting after you compile the firmware.

**Reply by danielnilsson9** [reply]
Can be set here before compiling: https://github.com/danielnilsson9/bbs-fw/blob/master/src/firmware/fwconfig.h#L107
Might be added to the config tool at some point in the future.

Quadratic is included but I tried it and it felt really bad since you basically lose 10% of your throttle range because the first 10% will be mapped to 0. Default map is x^1.5 which I found quite nice though.
